scripts/Diffraction/isis_powder/routines/InstrumentSettings.py

```diff
@@ -0,0 +1,108 @@
+from __future__ import (absolute_import, division, print_function)
+
+import warnings
+
+
+# Have to patch warnings at runtime to not print the source code. This is even advertised as a 'feature' of
+# the warnings library in the documentation: https://docs.python.org/3/library/warnings.html#warnings.showwarning
+def warning_no_source(msg, *ignored):
+    return str(msg) + '\n'
+
+warnings.formatwarning = warning_no_source
+warnings.simplefilter('always', UserWarning)
+
+
+class InstrumentSettings(object):
+    # Holds instance variables updated at runtime
+    def __init__(self, attr_mapping, adv_conf_dict=None, basic_conf_dict=None, kwargs=None):
+        self._attr_mapping = attr_mapping
+        self._adv_config_dict = adv_conf_dict
+        self._basic_conf_dict = basic_conf_dict
+        self._kwargs = kwargs
+
+        self._unknown_keys_found = False
+        self._parse_attributes(dict_to_parse=adv_conf_dict)
+        self._parse_attributes(dict_to_parse=basic_conf_dict)
+        self._parse_attributes(dict_to_parse=kwargs)
+        if self._unknown_keys_found:
+            _print_known_keys(attr_mapping)
+
+    def __getattr__(self, item):
+        map_entry = next((attr_tuple for attr_tuple in self._attr_mapping if item == attr_tuple[-1]), None)
+        if map_entry:
+            # User forgot to enter the param:
+            raise AttributeError("The parameter with name: '" + str(map_entry[0]) + "' is required but was not set or "
+                                 "passed.\nPlease set this configuration option and try again")
+        else:
+            # If you have got here from a grep or something similar this error message means the line caller
+            # has asked for a class attribute which does not exist. These attributes are set in a mapping file which
+            # is passed in whilst InstrumentSettings is being constructed. Check that the 'script name' (i.e. not user
+            # friendly name) is typed correctly in both the script(s) and mapping file.
+            raise AttributeError("The attribute in the script with name " + str(item) + " is unknown to the mapping."
```

Comment:
I'm reluctant to change this one - I've worded it slightly but I want them to contact the development team as it means a variable name internal to the script is incorrect.
Above that I have put a long comment explaining to future developers if they do see this message what it means and what they need to do.